### PR TITLE
RGW: Pass and check destination zone trace when fetching object

### PR DESCRIPTION
**Pull request by awojno-bloomberg**
The proposed fix for a race condition in multisite during full sync involving deletes. A custom header was added passing the destination zone's trace string and is then checked against the object's replication trace.  If there is a match, a 304 response is returned which will prevent full sync from overwriting a deleted object. Please see the [tracker](https://tracker.ceph.com/issues/58911) for more details.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by cbodley — COMMENTED** [review]
looks great!

File: src/rgw/rgw_sal.h [hunk 1]
```diff
@@ -884,6 +884,7 @@ class Object {
         const char* if_nomatch{nullptr};
         ceph::real_time* lastmod{nullptr};
         rgw_obj* target_obj{nullptr}; // XXX dang remove?
+        rgw_zone_set_entry* dst_zone_trace{nullptr};
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:887]: unused

Comment (awojno-bloomberg) [reply src/rgw/rgw_sal.h]: removed

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
@awojno-bloomberg looks like all of the required checks have passed 👍

**Comment by mkogan1** [comment]
> @[mattbenjamin](https://github.com/mattbenjamin) mattbenjamin requested a review from [mkogan1](https://github.com/mkogan1) [2 weeks ago](https://github.com/ceph/ceph/pull/51715#event-9328767259)

sorry still testing, the inconsistent state between zones keeps occuring with this PR on one test setup:
(but OK on other - still bisecting to find what is the difference between environments)

```
# main + this PR commit cherry-pick:
git branch -vv
* main                  2bc45cd9ef1 [origin/main: ahead 1] rgw: Pass and check destination zone trace when fetching object

❯ glo | grep -E 'rgw: Pass and check destination zone trace when fetching object|4023a15|9ec4fd7'
28ea361f867 (HEAD -> main) rgw: Pass and check destination zone trace when fetching object
9ec4fd7e497 rgw: fetch_remote_obj() adds x-rgw-replicated-from headers
4023a157add rgw/rados: pass source zone trace into fetch_remote_obj()
```
image
although the metadata and data are caught up, the object number in the data pool is not the same - 
left - primary zone, right secondary zone:

**Comment by awojno-bloomberg** [comment]
> image updating following more testing, that the _rados df_ stats mismatch is an intermittent temporary rados stats artifact which in above screenshot case corrected itself after 3 hours.
> 
> switched to `radosgw-admin bucket list ...` methodology for object count validation and result are immediate, consistent and correct.
> 
> LGTM

Great! Thanks for all the testing!

**Comment by mkogan1** [comment]
small addendum, found that the cause of the temporary  _rados df_ object count discrepancy was gc's objects, resolved by `radosgw-admin gc process --include-all`
image

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2023-06-08_19:52:02-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**